### PR TITLE
Fix for code containing only comments [closes #375]

### LINKED ISSUE #375: Parsing of comment-only files
Hi,

fparser2 is currently unable to parse comment-only files. I appreciate that it's debatable whether it should, however, I have one admittedly very specific use-case where I may end up splitting source-files into pieces before parsing them individually (each of those pieces valid and complete fortran program units -- or only comments). It's easy enough to work around this, of course, but I was wondering if you consider this to be something that should be working in general?

```
from pathlib import Path
fcode = """
! Some file
! that has only comments
! for whatever reason
""".strip()
Path('/tmp/myfile.F90').write_text(fcode)
```

```
from fparser.two.parser import ParserFactory
from fparser.common.readfortran import FortranFileReader
f2008_parser = ParserFactory().create(std="f2008")

reader = FortranFileReader("/tmp/myfile.F90", ignore_comments=False)
parse_tree = f2008_parser(reader)
print(parse_tree)
```

```
---------------------------------------------------------------------------
NoMatchError                              Traceback (most recent call last)
File ~/loki/nabr-lightweight-sourcefiles/loki_env/lib/python3.8/site-packages/fparser/two/Fortran2003.py:237, in Program.__new__(cls, string)
    236 try:
--> 237     return Base.__new__(cls, string)
    238 except NoMatchError:
    239     # At the moment there is no useful information provided by
    240     # NoMatchError so we pass on an empty string.

File ~/loki/nabr-lightweight-sourcefiles/loki_env/lib/python3.8/site-packages/fparser/two/utils.py:428, in Base.__new__(cls, string, parent_cls)
    427     errmsg = f"{cls.__name__}: '{string}'"
--> 428 raise NoMatchError(errmsg)

NoMatchError: at line 3
>>>! for whatever reason


During handling of the above exception, another exception occurred:

FortranSyntaxError                        Traceback (most recent call last)
Untitled-1.ipynb Cell 3 in <cell line: 2>()
      <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W1sdW50aXRsZWQ%3D?line=0'>1</a> reader = FortranFileReader("/tmp/myfile.F90", ignore_comments=False)
----> <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W1sdW50aXRsZWQ%3D?line=1'>2</a> parse_tree = f2008_parser(reader)
      <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W1sdW50aXRsZWQ%3D?line=2'>3</a> print(parse_tree)

File ~/loki/nabr-lightweight-sourcefiles/loki_env/lib/python3.8/site-packages/fparser/two/Fortran2003.py:241, in Program.__new__(cls, string)
    237     return Base.__new__(cls, string)
    238 except NoMatchError:
    239     # At the moment there is no useful information provided by
    240     # NoMatchError so we pass on an empty string.
--> 241     raise FortranSyntaxError(string, "")
    242 except InternalSyntaxError as excinfo:
    243     # InternalSyntaxError is used when a syntax error has been
    244     # found in a rule that does not have access to the reader
    245     # object. This is then re-raised here as a
    246     # FortranSyntaxError, adding the reader object (which
    247     # provides line number information).
    248     raise FortranSyntaxError(string, excinfo)

FortranSyntaxError: at line 3
>>>! for whatever reason
```

Just to demonstrate that this parses fine once I add some Fortran afterwards:

```
fcode += """
subroutine main
end
""".rstrip()
Path('/tmp/myotherfile.F90').write_text(fcode)
```
```
reader = FortranFileReader("/tmp/myotherfile.F90", ignore_comments=False)
parse_tree = f2008_parser(reader)
print(parse_tree)

! Some file
! that has only comments
! for whatever reason
SUBROUTINE main
END
```

FWIW: Out of curiosity I had a look at what compilers do with this: gfortran, ifort, ifx parse this without problems, nvfortran is the only one at least printing a warning stating that this file is empty. All generate `.o` files, though.

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for review.

File: src/fparser/two/Fortran2003.py [hunk 1]
```diff
@@ -12030,40 +12038,58 @@ def match(string):
...
     subclass_names = []
     use_names = ["Function_Name", "Dummy_Arg_Name_List", "Scalar_Expr"]
 
     @staticmethod
     def match(string):
-        i = string.find("=")
-        if i == -1:
-            return
-        expr = string[i + 1 :].lstrip()
+        """Implements the matching for an statement function statement.
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:12053]: "for a"

File: src/fparser/two/Fortran2003.py [hunk 2]
```diff
@@ -290,6 +290,9 @@ def match(reader):
         """
         content = []
         add_comments_includes_directives(content, reader)
+        comments = False
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:293]: `comments = (content != [])` perhaps?

File: src/fparser/two/Fortran2003.py [hunk 3]
```diff
@@ -12030,40 +12038,58 @@ def match(string):
...
-        expr = string[i + 1 :].lstrip()
+        """Implements the matching for an statement function statement.
+
+        :param str string: the string to match with as a statement
+            function statement.
+
+        :returns: a tuple of size 1 containing a Stmt_Function_Stmt \
+            object with the matched string if there is a match, or None \
+            if there is not.
+        :rtype: (:py:class:`fparser.two.Fortran2003.Stmt_Function_Stmt`) \
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:12061]: `Optional[Tuple[:py:class:...]]` please.

File: src/fparser/two/Fortran2003.py [hunk 4]
```diff
@@ -12030,40 +12038,58 @@ def match(string):
...
-        i = string.find("=")
-        if i == -1:
-            return
-        expr = string[i + 1 :].lstrip()
+        """Implements the matching for an statement function statement.
+
+        :param str string: the string to match with as a statement
+            function statement.
+
+        :returns: a tuple of size 1 containing a Stmt_Function_Stmt \
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:12058]: This method appears to return a 3-tuple. Please could you check and update the rtype too if necessary?

File: src/fparser/two/Fortran2003.py [hunk 5]
```diff
@@ -12030,40 +12038,58 @@ def match(string):
...
+            return None
+        index = line.find("(")
+        if index == -1:
+            return None
+        name = line[:index].rstrip()
         if not name:
-            return
-        args = line[i + 1 : -1].strip()
+            return None
+        args = line[index + 1 : -1].strip()
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:12080]: At this point you could check whether the arguments are in the symbol table. If they are then this isn't a statement function (I think).

File: src/fparser/two/Fortran2003.py [hunk 6]
```diff
@@ -12030,40 +12038,58 @@ def match(string):
...
         if args:
-            return Function_Name(name), Dummy_Arg_Name_List(args), Scalar_Expr(expr)
+            return (Function_Name(name), Dummy_Arg_Name_List(args), Scalar_Expr(expr))
         return Function_Name(name), None, Scalar_Expr(expr)
 
     def tostr(self):
+        """
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:12086]: You've defeated me - what was the issue with the statement function that you've fixed? (Maybe the tests will show me...)

**Review by arporter — CHANGES_REQUESTED** [review]
Fix looks fine. I can't work out what it was you've fixed in the Statement_Function code though.
Just a few minor things to tweak.
I got excited about using the Symbol Table to improve our matching of Statement Functions but that's really something for another Issue.

File: src/fparser/two/Fortran2003.py (hunk 2 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: Updated.

File: src/fparser/two/Fortran2003.py (hunk 1 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: stmt_function_stmt changes should not have been in this PR. I've removed them.

File: src/fparser/two/Fortran2003.py (hunk 3 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: stmt_function_stmt changes should not have been in this PR. I've removed them.

File: src/fparser/two/Fortran2003.py (hunk 4 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: stmt_function_stmt changes should not have been in this PR. I've removed them.

File: src/fparser/two/Fortran2003.py (hunk 5 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: stmt_function_stmt changes should not have been in this PR. I've removed them.

File: src/fparser/two/Fortran2003.py (hunk 6 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: stmt_function_stmt changes should not have been in this PR. I've removed them.

**Comment by rupertford** [comment]
The stmt_function_stmt changes were not menat to be in this PR. I've removed them.

**Comment by rupertford** [comment]
Read for next review.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
You could have kept the tidying of Statement_Function but that's OK.
Black is happy.
Proceeding to merge.

**Pull request merged**